pkg/controller/backup_queue_controller.go

```diff
@@ -0,0 +1,368 @@
+	backupNamespaces := namespacesForBackup(backup, clusterNamespaces)
+	backupNSMap := make(map[string]struct{})
+	for _, ns := range backupNamespaces {
+		backupNSMap[ns] = struct{}{}
+	}
+	for _, earlierBackup := range earlierBackups {
+		// This will never be true for the primary backup, but for the secondary
+		// runnability check for queued backups ahead of the current backup, we
+		// only care about backups ahead of it.
+		// Backup isn't earlier than this one, skip
+		if earlierBackup.Status.Phase == velerov1api.BackupPhaseQueued &&
+			earlierBackup.Status.QueuePosition >= backup.Status.QueuePosition {
+			continue
+		}
+		runningNSList := namespacesForBackup(&earlierBackup, clusterNamespaces)
+		for _, runningNS := range runningNSList {
+			if _, found := backupNSMap[runningNS]; found {
+				return true, earlierBackup.Name
+			}
+		}
+	}
+	return false, ""
+}
+
+// Returns true if there are backups ahead of the current backup that are runnable
+// This could happen if velero just reconciled the one earlier in the queue and rejected it
+// due to too many running backups, but a backup completed in between that reconcile and this one
+// so exit, as the recent completion has triggered another reconcile of all queued backups
+func (r *backupQueueReconciler) checkForEarlierRunnableBackups(backup *velerov1api.Backup, earlierBackups []velerov1api.Backup, clusterNamespaces []string) (bool, string) {
+	for _, earlierBackup := range earlierBackups {
+		// if this backup is queued and ahead of current backup, check for conflicts
+		if earlierBackup.Status.Phase != velerov1api.BackupPhaseQueued ||
+			earlierBackup.Status.QueuePosition >= backup.Status.QueuePosition {
+			continue
+		}
+		conflict, _ := detectNSConflictsInternal(&earlierBackup, earlierBackups, clusterNamespaces)
+		// !conflict means we've found an earlier backup that is currently runnable
+		// so current reconcile should exit to run this one
+		if !conflict {
+			return true, earlierBackup.Name
+		}
+	}
+	return false, ""
+}
+
+func namespacesForBackup(backup *velerov1api.Backup, clusterNamespaces []string) []string {
+	return collections.NewNamespaceIncludesExcludes().Includes(backup.Spec.IncludedNamespaces...).Excludes(backup.Spec.ExcludedNamespaces...).ActiveNamespaces(clusterNamespaces).ResolveNamespaceList()
+}
+func (r *backupQueueReconciler) getMaxQueuePosition(ctx context.Context, ns string) (int, error) {
+	queuedBackups, err := r.listQueuedBackups(ctx, ns)
+	if err != nil {
+		return 0, err
+	}
+	maxPos := 0
+	for _, backup := range queuedBackups {
+		maxPos = max(maxPos, backup.Status.QueuePosition)
+	}
+	return maxPos, nil
+}
+
+func (r *backupQueueReconciler) orderedQueuedBackups(ctx context.Context, backup *velerov1api.Backup) ([]velerov1api.Backup, error) {
+	backupList := &velerov1api.BackupList{}
+	var returnList []velerov1api.Backup
+	if err := r.Client.List(ctx, backupList, &client.ListOptions{Namespace: backup.Namespace}); err != nil {
```

Comment:
Why don't we list the backups in phase `BackupPhaseQueued` only and then do the sort?

Reply:
See https://github.com/vmware-tanzu/velero/pull/9307/files#r2521863068, we could have already cached the earlierBackups in `BackupPhaseQueued` phase, then we will not need to sort the entire backup list.

Reply:
I've refactored all of this now to only make one backup list call within Reconcile